storage/innobase/buf/buf0buf.cc

```diff
@@ -1446,6 +1454,206 @@ buf_block_init(
+	return true;
+#else  /* HAVE_MADV_DONTDUMP */
+	ib::warn() << "Disabling @@core_file because "
+		      "@@innodb_buffer_pool_in_core_file is disabled, yet "
+		      "MADV_DONTDUMP is not supported on this platform";
+	return false;
+#endif /* HAVE_MADV_DONTDUMP */
+}
+
+/* End of implementation of buf_chunk_t's methods */
+
+/* Implementation of buf_pool_t's methods */
+
+/** A wrapper for buf_pool_t::allocator.alocate_large which also advices the OS
+that this chunk should not be dumped to a core file if that was requested.
+Emits a warning to the log and disables @@global.core_file if advising was
+requested but could not be performed, but still return true as the allocation
+itself succeeded.
+@param[in]	mem_size  number of bytes to allocate
+@param[in/out]  chunk     mem and mem_pfx fields of this chunk will be updated
+                          to contain information about allocated memory region
+@return true iff allocated successfully */
+bool
+buf_pool_t::allocate_chunk(ulonglong mem_size, buf_chunk_t *chunk,
+			   bool populate)
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	chunk->mem =
+	    allocator.allocate_large(mem_size, &chunk->mem_pfx, populate);
+	if (UNIV_UNLIKELY(chunk->mem == NULL)) {
+		return false;
+	}
+	/* Dump core without large memory buffers */
+	if (buf_pool_should_madvise) {
+		if (!chunk->madvise_dont_dump()) {
+			innobase_disable_core_dump();
+		}
+	}
+	return true;
+}
+
+/** A wrapper for buf_pool_t::allocator.deallocate_large which also advices the
+OS that this chunk can be dumped to a core file.
+Emits a warning to the log and disables @@global.core_file if advising was
+requested but could not be performed.
+@param[in]  chunk    mem and mem_pfx fields of this chunk will be used to locate
+                     the memory region to free */
+void
+buf_pool_t::deallocate_chunk(buf_chunk_t *chunk)
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	/* Undo the effect of the earlier MADV_DONTDUMP */
+	if (buf_pool_should_madvise) {
+		if (!chunk->madvise_dump()) {
+			innobase_disable_core_dump();
+		}
+	}
+	allocator.deallocate_large(chunk->mem, &chunk->mem_pfx);
+}
+
+/** Advices the OS that all chunks in this buffer pool instance can be dumped
+to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
+bool
+buf_pool_t::madvise_dump()
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	for (buf_chunk_t *chunk = chunks; chunk < chunks + n_chunks;
+	     chunk++) {
+		if (!chunk->madvise_dump()) {
+			return false;
+		}
+	}
+	return true;
+}
+
+/** Advices the OS that all chunks in this buffer pool instance should not
+be dumped to a core file.
+Emits a warning to the log if could not succeed.
+@return true iff succeeded, false if no OS support or failed */
+bool
+buf_pool_t::madvise_dont_dump()
+{
+	ut_ad(mutex_own(&chunks_mutex));
+	for (buf_chunk_t *chunk = chunks; chunk < chunks + n_chunks;
+	     chunk++) {
+		if (!chunk->madvise_dont_dump()) {
+			return false;
```

Comment:
same here.. convert all return without braces to return(X); style.. (this is old 5.7 style)